### PR TITLE
[Proposal] Reduce internal dependancy

### PR DESCRIPTION
**Pull request by Chachay**
- Utilize scipy module to reduce internal dependency (cubic spline)
- Reduce high context parameters in rear_wheel_feedback_control function

Other proposal (not implemented): 
- Speed profile is disabled; could direction switching be decided at line 143? this would be closer to real-world implementation.
- pi_2_pi could also be replaced to numpy unwrap; though I think it is not so effective on readability.

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
@Chachay Thank you for your PR!!. It looks simpler. Please fix some minor points to merge as below.

**Comment by AtsushiSakai** [comment]
Please fix the issues reported by lgtm-com.

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 1]
```diff
@@ -26,34 +22,79 @@
...
+
+        self.dX = self.X.derivative(1)
+        self.ddX = self.X.derivative(2)
+
+        self.dY = self.Y.derivative(1)
+        self.ddY = self.Y.derivative(2)
+
+        self.length = s[-1]
+    
+    def yaw(self, s):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: calc_yaw is better? function should be start by a verb.

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 2]
```diff
@@ -26,34 +22,79 @@
...
+        dx, dy = self.dX(s), self.dY(s)
+        return np.arctan2(dy, dx)
+    
+    def curvature(self, s):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: calc_curvature?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 3]
```diff
@@ -26,34 +22,79 @@
...
+        dx, dy   = self.dX(s), self.dY(s)
+        ddx, ddy   = self.ddX(s), self.ddY(s)
+        return (ddy * dx - ddx * dy) / ((dx ** 2 + dy ** 2)**(3 / 2))
+    
+    def __findClosestPoint(self, s0, x, y):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Please use snake-case based on pep8.

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 4]
```diff
@@ -26,34 +22,79 @@
...
+        def f(_s, *args):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Can you make the function name more explicit?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 5]
```diff
@@ -26,34 +22,79 @@
...
+            _x, _y= self.X(_s), self.Y(_s)
+            return (_x - args[0])**2 + (_y - args[1])**2
+        
+        def jac(_s, *args):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Can you make the function name more explicit?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 6]
```diff
@@ -26,34 +22,79 @@
...
+            _x, _y = self.X(_s), self.Y(_s)
+            _dx, _dy = self.dX(_s), self.dY(_s)
+            return 2*_dx*(_x - args[0])+2*_dy*(_y-args[1])
+
+        minimum = optimize.fmin_cg(f, s0, jac, args=(x, y), full_output=True, disp=False)
+        return minimum
+
+    def TrackError(self, x, y, s0):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Should be calc_track_error

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 7]
```diff
@@ -26,34 +22,79 @@
...
+
+        dxl = self.X(s) - x
+        dyl = self.Y(s) - y
+        angle = pi_2_pi(yaw - math.atan2(dyl, dxl))
+        if angle < 0:
+            e*= -1
+
+        return e, k, yaw, s
 
 def PIDControl(target, current):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Please change pid_control

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 8]
```diff
@@ -26,34 +22,79 @@
...
-
-    return state
-
+    def update(self, a, delta, dt):
+        self.x   = self.x + self.v * math.cos(self.yaw) * dt
+        self.y   = self.y + self.v * math.sin(self.yaw) * dt
+        self.yaw = self.yaw + self.v / L * math.tan(delta) * dt
+        self.v   = self.v + a * dt
+
+class TrackSpline:
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: CubicSplinePath is better?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 1 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: renamed

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 2 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: renamed

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 3 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: renamed

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 5 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: named calc distance & its jacobian

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 7 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: renamed

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 8 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: changed

**Comment by Chachay** [comment]
Thank you for your review. I added "switching direction" in addition to the function name fixes.

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 9]
```diff
@@ -147,64 +160,46 @@ def closed_loop_prediction(cx, cy, cyaw, ck, speed_profile, goal):
...
     ax = [0.0, 6.0, 12.5, 5.0, 7.5, 3.0, -1.0]
     ay = [0.0, 0.0, 5.0, 6.5, 3.0, 5.0, -2.0]
     goal = [ax[-1], ay[-1]]
 
-    cx, cy, cyaw, ck, s = cubic_spline_planner.calc_spline_course(
-        ax, ay, ds=0.1)
-    target_speed = 10.0 / 3.6
-
-    sp = calc_speed_profile(cx, cy, cyaw, target_speed)
+    track = CubicSplinePath(ax, ay)
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: track -> cubic_spline_path ?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py [hunk 10]
```diff
@@ -63,53 +104,24 @@ def pi_2_pi(angle):
 
     return angle
 
-
-def rear_wheel_feedback_control(state, cx, cy, cyaw, ck, preind):
-    ind, e = calc_nearest_index(state, cx, cy, cyaw)
-
-    k = ck[ind]
+def rear_wheel_feedback_control(state, e, k, yaw_r):
```
Comment (AtsushiSakai) [inline PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: yaw_r -> yaw_ref ?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 9 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: couldn't this object be better to have the name detached from the interpolation method?

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 10 above)
Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: I will rename this

File: PathTracking/rear_wheel_feedback/rear_wheel_feedback.py (hunk 9 above)
Comment (AtsushiSakai) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: Does this object means a spline path object? What do you imply the "track" name?

Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: I meant the name could be any of reference_track, reference_path, target_track and target_path or even only path or track, but it isn't necessary to be cubic spline as long as class C2.

Comment (AtsushiSakai) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: oh ok. I got it. i feel reference_path is better.

Comment (Chachay) [reply PathTracking/rear_wheel_feedback/rear_wheel_feedback.py]: changed.

**Pull request merged**